man/natspec.1
'\" t
.\"     Title: NATSPEC
.\"    Author: Andrew Shadoura <bugzilla@tut.by>
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 09/05/2009
.\"    Manual: natspec User Manual
.\"    Source: natspec
.\"  Language: English
.\"
.TH "NATSPEC" "1" "09/05/2009" "natspec" "natspec User Manual"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
natspec \- command line utility to access libnatspec functions
.SH "SYNOPSIS"
.HP \w'\fBnatspec\fR\ 'u
\fBnatspec\fR [\fB\-c\fR\ |\ \fB\-\-codepage\fR] [\fB\-f\fR\ |\ \fB\-\-fsenc\fR] [\fB\-\-utf8\fR] [\fB\-\-country\-id\fR] [\fB\-\-system\-locale\fR] [\fB\-i\fR\ |\ \fB\-\-info\fR] [\fB\-a\ \fR\fB\fIencoding\fR\fR\ |\ \fB\-\-transl\ \fR\fB\fIencoding\fR\fR] [\fB\-v\fR\ |\ \fB\-\-verbose\fR] [\fB\-s\ \fR\fB\fIoperating\ system\fR\fR\ |\ \fB\-\-charset\ \fR\fB\fIoperating\ system\fR\fR]
.HP \w'\fBnatspec\fR\ 'u
\fBnatspec\fR [\fB\-h\fR\ |\ \fB\-\-help\fR | \fB\-V\fR\ |\ \fB\-\-version\fR]
.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBnatspec\fR
command\&.
.PP
This manual page was written for the Debian distribution because the original program does not have a manual page\&.
.PP
\fBnatspec\fR
is a command line utility to access libnatspec functions\&.
.SH "OPTIONS"
.PP
The program follows the usual GNU command line syntax, with long options starting with two dashes ("\-")\&. A summary of options is included below\&.
.PP
\fB\-c\fR, \fB\-\-codepage\fR
.RS 4
Print DOS codepage for current locale\&.
.RE
.PP
\fB\-f\fR, \fB\-\-fsenc\fR
.RS 4
Print filename encoding for current locale\&.
.RE
.PP
\fB\-\-utf8\fR
.RS 4
Return 0 if current locale uses UTF\-8\&.
.RE
.PP
\fB\-\-country\-id\fR
.RS 4
Print country ID for DOS\&.
.RE
.PP
\fB\-\-system\-locale\fR
.RS 4
Print system locale\&.
.RE
.PP
\fB\-i\fR, \fB\-\-info\fR
.RS 4
Print summary information about current locale\&.
.RE
.PP
\fB\-a \fR\fB\fIencoding\fR\fR, \fB\-\-transl=\fR\fB\fIencoding\fR\fR
.RS 4
Transliterate
\fIARG\fR
from
\fIencoding\fR
to the current encoding and print the result\&.
.RE
.PP
\fB\-v\fR, \fB\-\-verbose\fR
.RS 4
Be verbose\&.
.RE
.PP
\fB\-s \fR\fB\fIoperating system\fR\fR, \fB\-\-charset=\fR\fB\fIoperating system\fR\fR
.RS 4
Print charset for the
\fIoperating system\fR\&.
\fIoperating system\fR
can be one of unix, win, dos, mac\&.
.RE
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Show summary of options\&.
.RE
.PP
\fB\-V\fR, \fB\-\-version\fR
.RS 4
Show version of program\&.
.RE
.PP
When being called without options,
\fB\-i\fR
is implied\&.
.SH "ENVIRONMENT"
.PP
\fBLC_ALL\fR
.RS 4
Used for detecting current locale\&.
.RE
.SH "AUTHOR"
.PP
\fBAndrew Shadoura\fR <\&bugzilla@tut\&.by\&>
.RS 4
Wrote this manpage for the Debian system\&.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2009 Andrew O. Shadoura
.br
.PP
This manual page was written for the Debian system (and may be used by others)\&.
.PP
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU Lesser General Public License, Version 2\&.1 or (at your option) any later version published by the Free Software Foundation\&.
.PP
On Debian systems, the complete text of the GNU Lesser General Public License can be found in
/usr/share/common\-licenses/LGPL\-2\&.1\&.
.sp


NEWS
libnatspec-0.2.4: 15-Feb-2008:
 - natspec_convert returns original string if failed (fix bug #14301)
 - small bugfixes, clean code

libnatspec-0.2.3: 02-Jan-2008:
 - add patches for unzip, pmount
 - fix build

libnatspec-0.2.3-beta: 13-Mar-2006:
 - add natspec_iconv, natspec_iconv_open, natspec-iconv_close
 - add transliterate param to natspec_convert (set to 0 for old behaviour)
 - obsolete natspec_convert_with_translit (use natspec_convert with last param set to 1)
 - add --system-locale in natspec
 - add patch for o3read
 - add patch for catpkt from Igor Zubkov <icesik@mail.ru>
 - check G_FILENAME_ENCODING for @locale value
 - add patch for dict client, isoinfo from cdrtools
 
libnatspec-0.2.2: 22-Jul-2005:
 - fix use static buffer natspec_get_codepage (change type of returned value to int)
 - review code and some cleanup
 - introduce natspec_locale_is_utf8 function
 - do not use LANGUAGE env variable
 - build API documentation

libnatspec-0.2.1: 09-Apr-2005:
 - fix CP1125 order
 - fix locale string parsing
 - add print current charset in natspec util (in info mode)

libnatspec-0.2.0: 17-Mar-2005:
 - rename get_user_locale() to get_current_locale()
 - discard from toupper/tolower/isalnum/strcasecmp using (it was locale depended)
 - fix DOS charset for uk_UA
 - introduce samba/loadparams (for server autotuning) - does not work yet

libnatspec-0.1.9: 03-Mar-2005:
 - add jfs support
 - update submount patch (to be simple)
 - update samba patch (add default charsets for params)
 - fixes for compilation in various OS
 - natspec: print overall information by default, print ASCII for unknown system
 - do not use setlocale anywhere
 - add --with-i18n-file option for set file with LANG variable at your system
 - add --with-python option to configure
 - improved get_filesystem_encoding
 - humble_charset do toupper now and breaks at ':' symbol
 - repack_locale do tolower for lang and toupper for country string

libnatspec-0.1.0: 27-Feb-2005:
 - fixed locale using order, add some fallbacks for unusual situation
 - update docs

libnatspec-0.0.9: 27-Feb-2005:
 - rename enrich_fs_options to get_enriched_fs_options with incompatible change args
 - update mount/submount, sdcv example patches
 - add example path for samba (smbmount/cifsmount)
 - add auto assumed local encoding in natspec_convert
 - add convert_with_translit with transliteration
   (from links project) if conversion is not possible
 - add natspec.m4 for autotools, it makes natspec configure is easy

libnatspec-0.0.4: 24-Feb-2005:
 - add python binding
 - code cleanup
 - cleanup comments in header file
 - add README with short description (in English)
 - add README-ru.html with description (in Russian)
 
libnatspec-0.0.3: 23-Feb-2005:
 - fix bug with unix charset (was due misunderstanging wine behaviour)
 - add examples of patch for mount, submount, mkisofs, isoinfo, stardict-sdcv
 
libnatspec-0.0.2: 22-Feb-2005:
 - add get_charset function
 - add convert for lossy conversion between encoding
 - add smbfs support in enrich_fs_options

libnatspec-0.0.1: 20-Feb-2005:
 - first public release after born
 - support for mount filesystem (enrich_fs_options)
 - system filename encoding detection (get_filename_encoding)
 - get system locale from some file (/etc/sysconfig/i18n) (get_system_locale)
 - retrieve charset/codepage for another OS by system/current
   locale/charset (get_charset_by_locale)
 - command line util for using features in shells (natspec)


src/natspec.c
/*
    src/natspec.c

    command line util for access to libnatspec functions

    NATSPEC library
    Copyright (c) 2005, 2015 Etersoft
    Copyright (c) 2005, 2015 Vitaly Lipatov <lav@etersoft.ru>

    $Id: natspec.c,v 1.22 2005/09/03 18:59:28 vitlav Exp $

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation, version 2.1
    of the License.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
    USA.
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <strings.h>
#include <assert.h>
#include <locale.h>

#include "natspec_internal.h"
#include "natspechi_internal.h"

#ifdef HAVE_LIBPOPT
#  include <popt.h>
#endif

#include <locale.h>
#ifdef HAVE_LANGINFO_H
# include <langinfo.h>
#endif

#define VERBOSE if (verbose) printf

/* natspec internals */
char *natspec_internal_get_locale_from_env();

char *charset_type;
char *locale, *transliterate;
int version, verbose, fsenc, fcodepage, nls, flag_help, info;
int utf8, country_id, get_system_locale, get_user_locale;

#if defined HAVE_LIBPOPT
poptContext context = NULL;
struct poptOption options[] =
{
    {"codepage", 'c', POPT_ARG_NONE,  &fcodepage, 1,
     "print codepage", ""},
    {"fsenc", 'f', POPT_ARG_NONE,  &fsenc, 1,
     "print filesystem encoding", ""},
    {"utf8", 0, POPT_ARG_NONE,  &utf8, 1,
     "return 0 if locale is utfed", ""},
    {"country-id", 0, POPT_ARG_NONE,  &country_id, 1,
     "print country ID (international numbering)", ""},
    {"locale", 'l', POPT_ARG_NONE,  &get_user_locale, 1,
     "print current user locale", ""},
    {"system-locale", 0, POPT_ARG_NONE,  &get_system_locale, 1,
     "print system locale", ""},
    {"info", 'i', POPT_ARG_NONE,  &info, 0,
     "print overall encoding/charset info for your system", ""},
    {"transl", 'a', POPT_ARG_STRING,  &transliterate, 1,
     "transliterate arg from ENC to current encoding and print it", ""},
    {"version", 'V', POPT_ARG_NONE, &version, 1,
     "display version and exit", NULL },
    {"verbose", 'v', POPT_ARG_NONE, &verbose, 0,
     "verbose output", NULL },
    {"charset", 's', POPT_ARG_STRING,  &charset_type, 0,
     "print charset for op. system: unix, win, dos, mac (case insensivity)", ""},
    {"help", 'h', POPT_ARG_NONE, &flag_help, 1, "Show this help message" },
    {(char *) NULL, '\0', 0, NULL, 0}
};
#endif

static const char *get_charset(char *charset_type)
{
	/* FIXME: strcasecmp is locale depends */
	const char *charset;
	int type;
	assert (charset_type);
	if (!strcasecmp(charset_type,"WIN"))
		type = NATSPEC_WINCS;
	else if (!strcasecmp(charset_type,"DOS"))
		type = NATSPEC_DOSCS;
	else if (!strcasecmp(charset_type,"MAC"))
		type = NATSPEC_MACCS;
	else if (!strcasecmp(charset_type,"UNIX"))
		type = NATSPEC_UNIXCS;
	else
	{
		VERBOSE("We do not know %s type of encoding\n",charset_type);
		printf("ASCII\n");
		exit(1);
	}
	charset = natspec_get_charset_by_locale(type, locale);
	VERBOSE("Charset of '%s' system: ", charset_type);
	printf("%s\n", charset);
	return charset;
}


int main(int argc, const char** argv)
{
	int is_utf8;
#if defined HAVE_LIBPOPT
	int rc = 0;
	poptContext poptCtx;
	poptCtx = poptGetContext("natspec", argc, (const char **)argv, options, 0);
	poptSetOtherOptionHelp(poptCtx, "[OPTION...]");
	while (rc >= 0) {
		if((rc = poptGetNextOpt(poptCtx)) < -1) {
			fprintf(stderr, "Error on option %s: %s.\nRun '%s --help' to see a full list of available command line options.\n",
				poptBadOption(poptCtx, 0),
				poptStrerror(rc),
				argv[0]);
			exit(1);
		}
	}
/*
  { int flag_help=1;
  if (argv) {
      while (poptPeekArg(poptCtx))
		flag_help=0;
		//locale = (char *)poptGetArg(poptCtx);
      poptFreeContext(poptCtx);
	}
*/	
	if(flag_help)
	{
		poptPrintHelp(poptCtx, stdout, 0);
		exit(0);
	}

#else
	fprintf(stderr, "Compiled without popt. Exit\n");
	exit(1);
#endif

        /* Print overall information by default */
	if (argc == 1)
		info = 1;

	if (version || info)
	{
		printf("%s, compiled %s\n", PACKAGE_STRING,__DATE__);
		if (!info)
			exit(0);
	}
	if (transliterate)
	{
		char *str = NULL;
		if (argv) {
#if defined HAVE_LIBPOPT
			while (poptPeekArg(poptCtx))
				str = (char *)poptGetArg(poptCtx);
			poptFreeContext(poptCtx);
#else
			fprintf(stderr, "Compiled without popt. Exit\n");
			exit(1);
#endif
		}
		if ( str != NULL)
		{
			setlocale(LC_ALL,"");
			str = natspec_convert(str, "", transliterate, 1);
			puts(str);
		}
		exit(0);
	}
	if (info)
	{
		printf(" === Overall locale information ===\n");
		verbose = 1;
	}
	locale = natspec_get_current_locale();
	if (get_user_locale || info)
	{
		VERBOSE("Current user locale: ");
		printf("%s", locale);
		if (verbose && !natspec_internal_get_locale_from_env())
			printf(" (got from system locale)");
		puts("");
	}
	if (get_system_locale || info)
	{
		VERBOSE("System locale: ");
		printf("%s\n", natspec_get_system_locale());
		if (!info)
			exit(0);
	}
	if (fsenc || info)
	{
		const char *buf;
		VERBOSE("Filename encoding in iconv/nls form: ");
		buf = natspec_get_filename_encoding("");
		printf("%s\n",buf);
		if (!info)
			exit(0);
	}
	if (country_id || info)
	{
		int id;
		VERBOSE("Country ID (for DOS): ");
		id = -1; /* natspec_get_country_id(); */
		printf("%d\n", id);
		if (!info)
			exit(0);
	}
	if (fcodepage || info)
	{
		int cp;
		VERBOSE("Codepage of DOS: ");
		cp = natspec_get_codepage_by_charset(
			natspec_get_charset_by_locale(NATSPEC_DOSCS, locale));
		printf("%d\n", cp);
		if (!info)
			exit(0);
	}

	if (info)
	{
		char *types[]={"UNIX","WIN","DOS","MAC"};
		int i;
		for (i=0; i<sizeof(types)/sizeof(char*);i++)
			get_charset(types[i]);
		setlocale(LC_ALL,"");
		printf("After setlocale:\n");
#ifdef HAVE_LANGINFO_H
		printf("\tnl_langinfo(CODESET): %s\n",nl_langinfo(CODESET));
#endif
		printf("\tnatspec_get_charset: %s\n",natspec_get_charset());
	}

	if (charset_type)
		get_charset(charset_type);
	is_utf8 = natspec_locale_is_utf8("");

	if (utf8 || info)
	{
		VERBOSE("Current locale is%sin UTF8 encoding\n",(is_utf8 ? " " : " NOT "));
		if (!info)
			return is_utf8 ? 0 : 1;
	}
	return 0;
}
